include/gsl/gsl-lite.hpp

```diff
@@ -1372,26 +1433,75 @@ class not_null
 # endif
 #endif
 
-    template< class U
-#if gsl_HAVE( DEFAULT_FUNCTION_TEMPLATE_ARG )
-        gsl_REQUIRES_T(( std::is_convertible<U, T>::value ))
+    // converting copy constructors and assignment from not_null<U>:
+    // without type_traits, we can't distinguish is_convertible and is_constructible, so all converting constructors are explicit
+
+#if gsl_HAVE( TYPE_TRAITS )
+    template< class U >
+    gsl_api gsl_constexpr not_null( not_null<U> const & other gsl_REQUIRES_A((std::is_convertible<U, T>::value)) )
+    : ptr_( other.checked_ptr_ref() )
+    {}
 #endif
-    >
-    gsl_api gsl_constexpr not_null( not_null<U> const & other )
-    : ptr_( other.get() )
+
+    template< class U >
+    gsl_api gsl_constexpr explicit not_null( not_null<U> const & other
+#if gsl_HAVE( TYPE_TRAITS )
+        gsl_REQUIRES_A((std::is_constructible<T, U>::value && !std::is_convertible<U, T>::value))
+#endif
+    )
+    : ptr_( other.checked_ptr_ref() )
+    {}
+
+    template< class U >
+    gsl_api gsl_constexpr14 not_null<T>& operator=( not_null<U> const & other)
```

Comment:
I once had some problems with a compiler (I think qcc 4.7.3, a fork of gcc for the QNX platform) with this, so I'd prefer to leave it this way. Also, I think it's clearer this way, since there's already another not_null in this line.

Reply:
OK.